syntax = "proto3";
option java_multiple_files = true;
package palm.plugins.nut.v1;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

message Pager {
  int64 page = 1;
  int64 size = 2;
}

message Pagination {
  int64 page = 1;
  int64 size = 2;
  int64 total = 3;
  bool has_next = 11;
  bool has_previous = 12;
}

message IdRequest {
  int32 id = 1;
}

service Locale {
  rpc Set(LocaleSetRequest) returns (google.protobuf.Empty) {}
  rpc Get(LocaleGetRequest) returns (LocaleIndexResponse.Item) {}
  rpc Index(Pager) returns (LocaleIndexResponse) {}
  rpc Destroy(IdRequest) returns (google.protobuf.Empty) {}
}

message LocaleIndexResponse {
  message Item {
    int32 id = 1;
    string lang = 2;
    string code = 3;
    string message = 4;
    google.protobuf.Timestamp updated_at = 9;
  }
  repeated Item items = 1;
  Pagination pagination = 9;
}

message LocaleGetRequest {
  string lang = 1;
  string code = 2;
}
message LocaleSetRequest {
  string lang = 1;
  string code = 2;
  string message = 3;
}

// ----------------------------------------------------------------------------

service Site {
  rpc ClearCache(google.protobuf.Empty) returns (google.protobuf.Empty) {}

  rpc Layout(google.protobuf.Empty) returns (SiteLayoutResponse) {}
  rpc SetAuthor(SiteLayoutResponse.Author) returns (google.protobuf.Empty) {}
  rpc SetCopyright(SiteSetCopyrightRequest) returns (google.protobuf.Empty) {}
  rpc SetKeywords(SiteSetKeywordsRequest) returns (google.protobuf.Empty) {}
  rpc SetInfo(SiteSetInfoRequest) returns (google.protobuf.Empty) {}
  rpc SetLogo(SiteSetLogoRequest) returns (google.protobuf.Empty) {}

  rpc SetAws(AwsProfile) returns (google.protobuf.Empty) {}
  rpc GetAws(google.protobuf.Empty) returns (AwsProfile) {}
  rpc TestAwsS3(google.protobuf.Empty) returns (SiteAwsS3TestResponse) {}
  rpc SetSmtp(SmtpProfile) returns (google.protobuf.Empty) {}
  rpc GetSmtp(google.protobuf.Empty) returns (SmtpProfile) {}
  rpc TestSmtp(SiteSmtpTestRequst) returns (google.protobuf.Empty) {}
  rpc SetBing(BingProfile) returns (google.protobuf.Empty) {}
  rpc GetBing(google.protobuf.Empty) returns (BingProfile) {}
  rpc SetGoogle(GoogleProfile) returns (google.protobuf.Empty) {}
  rpc GetGoogle(google.protobuf.Empty) returns (GoogleProfile) {}
  rpc SetBaidu(BaiduProfile) returns (google.protobuf.Empty) {}
  rpc GetBaidu(google.protobuf.Empty) returns (BaiduProfile) {}

  rpc Status(google.protobuf.Empty) returns (SiteStatusResponse) {}

  rpc NewLeaveWord(SiteNewLeaveWordRequest) returns (google.protobuf.Empty) {}
  rpc IndexLeaveWord(Pager) returns (SiteIndexLeaveWordResponse) {}
  rpc DestroyLeaveWord(IdRequest) returns (google.protobuf.Empty) {}
}

message AwsProfile {
  string region = 1;
  optional string endpoint = 2;
  string access_key_id = 3;
  string secret_access_key = 4;
}
message SiteAwsS3TestResponse {
  repeated string buckets = 1;
}

message SiteSmtpTestRequst {
  string to = 1;
  string subject = 2;
  string body = 3;
}

message SiteSetLogoRequest {
  string url = 1;
}
message SiteSetCopyrightRequest {
  string payload = 1;
}
message SiteSetKeywordsRequest {
  repeated string items = 1;
}
message SiteSetInfoRequest {
  string title = 1;
  string subhead = 2;
  string description = 3;
}

message SiteLayoutResponse {
  message Author {
    string name = 1;
    string email = 2;
  }
  string title = 1;
  string subhead = 2;
  repeated string keywords = 3;
  string description = 4;
  Author author = 5;
  string copyright = 6;
  string logo = 7;
  repeated string languages = 9;
}

message SiteNewLeaveWordRequest {
  string body = 1;
}
message SiteIndexLeaveWordResponse {
  message Item {
    int32 id = 1;
    string lang = 2;
    string ip = 3;
    string body = 4;
    google.protobuf.Timestamp created_at = 9;
  }
  repeated Item items = 1;
  Pagination pagination = 9;
}

message GoogleProfile {
  // https://developers.google.com/recaptcha/intro
  message ReCaptcha {
    string site_key = 1;
    string secret = 2;
  }
  optional string site_verify_id = 1;
  optional ReCaptcha re_captcha = 2;
}
message BaiduProfile {
  message SiteVerify {
    string id = 1;
    string content = 2;
  }
  optional SiteVerify site_verify = 1;
}
message BingProfile {
  optional string site_verify_id = 1;
}

message EmailTask {
  enum ContentType {
    TEXT_PLAIN = 0;
    TEXT_HTML = 1;
  }
  message Attachment {
    string name = 1;
    ContentType content_type = 2;
    bytes payload = 11;
  }
  string subject = 1;
  string body = 2;
  ContentType content_type = 3;

  string to = 11;
  repeated string cc = 12;
  repeated string bcc = 13;
  repeated Attachment attachments = 21;
}

message SmtpProfile {
  string host = 1;
  uint32 port = 2;
  string user = 3;
  string password = 4;
  string cc = 5;
  string bcc = 6;
}

message SiteStatusResponse {
  message Database {
    string name = 1;
    string size = 2;
  }
  message PostgreSql {
    string version = 1;
    google.protobuf.Timestamp now = 2;
    repeated Database databases = 9;
  }
  message MySql {
    uint64 size = 1;
    string version = 2;
  }
  message Redis {
    message Item {
      string key = 1;
      int64 ttl = 2;
    };
    string info = 1;
    repeated Item items = 2;
  }
  message RabbitMq {
    string protocol = 1;
  }
  message OpenSearch {
    string url = 1;
    string info = 2;
  }
  message Health {
    string name = 1;
    string status = 2;
  }
  message System {
    string version = 1;
    string cpu = 2;
    string memory = 3;
    string boot = 4;
    string disk = 5;
    string load = 6;
    string fs = 7;
    string swap = 8;
    string uptime = 9;
    string network = 10;
  }

  PostgreSql postgresql = 1;
  MySql mysql = 2;
  Redis redis = 3;
  RabbitMq rabbitmq = 4;
  OpenSearch opensearch = 5;
  System system = 11;
  repeated Health healthes = 21;
}
